src/nimony/semdecls.nim

```diff
@@ -654,8 +654,119 @@ proc semProcImpl(c: var SemContext; it: var Item; kind: SymKind; pass: PassKind;
     producesVoid c, info, it.typ
   publish c, symId, declStart
 
+proc findMacroInvocs(c: SemContext; n: Cursor; kind: SymKind): seq[Cursor] =
+  # find all macro/template identifiers in pragmas to invoke them with parent proc definition
+  result = newSeq[Cursor]()
+  if kind in {ProcY, FuncY}:
+    var n = asRoutine(n).pragmas
+    if n.substructureKind == PragmasU:
+      inc n
+      while n.kind != ParRi:
+        if n.exprKind == ErrX or n.substructureKind == KvU:
+          skip n
+        elif pragmaKind(n) != NoPragma or callConvKind(n) != NoCallConv:
+          skip n
+        else:
+          let hasParRi = n.kind == ParLe
+          let start = n
+          if n.exprKind == CallX:
+            inc n
+          let name = getIdent(n)
+          if name != StrId(0) and not (name in c.userPragmas and not hasParRi):
+            result.add start
+            n = start
+            skip n
+          else:
+            skip n
+
+proc addCall(buf: var TokenBuf;
+             bufReadPos: var seq[int];
+             i: int;
+             macroInvocsPos: seq[Cursor];
+             info: PackedLineInfo) =
+  # adds CallX to invoke all macros/templates at first
+  # to avoid inserting them on the top of buf after the invocations.
+  bufReadPos.add buf.len
+  buf.addParLe CallX, info
+  var n = macroInvocsPos[^i]
+  let isCall = n.exprKind == CallX
+  if isCall:
+    inc n
+  assert n.kind == Ident
+  buf.add n
+  if isCall:
+    inc n
+    while n.kind != ParRi:
+      buf.takeTree n
+  buf.addParLe StmtsS, info
+
+proc transformMacroInvoc(c: var SemContext; it: var Item; macroInvocsPos: seq[Cursor]) =
+  # transform `proc foo() {.macrofoo, macrobar.}` to `macrobar: macrofoo: proc foo()`.
+  var inBuf = createTokenBuf()
+  var outBuf = createTokenBuf()
+  var inBufReadPos = newSeqOfCap[int](macroInvocsPos.len div 2)
+  var outBufReadPos = newSeqOfCap[int](macroInvocsPos.len div 2)
+
+  # adds last one in macroInvocsPos to buf first as it is invoked last.
+  let info = it.n.info
+  var offset = 0
+  if macroInvocsPos.len mod 2 == 1:
+    addCall inBuf, inBufReadPos, 1, macroInvocsPos, info
+    offset = 1
+  for i in 0 ..< (macroInvocsPos.len div 2):
+    addCall outBuf, outBufReadPos, i * 2 + 1 + offset, macroInvocsPos, info
+    addCall  inBuf,  inBufReadPos, i * 2 + 2 + offset, macroInvocsPos, info
```

Comment:
What is this? Rewrite this, all you need to do is to create a call expression, `templateHere: proc ...` that you semcheck. What problem does this complex logic solve? Why are there outBuf and inBuf that both get some calls?